### PR TITLE
feat(agw): Run integ tests vs containerized AGW

### LINKED ISSUE #13475: Run integration test against containerized AGW
## Acceptance criteria
- [x] Deploy the containerized AGW following the [documentation](https://github.com/magma/magma/blob/master/lte/gateway/docker/README.md#running-the-containerized-agw-locally-on-the-magma-vm).
- [x] Run AGW integration tests against the containerized AGW. Get test_attach_detach to run.
- [x] (Optional) Run more tests. Ones involving the traffic server would be a bonus.
- [x] Document
  - the necessary steps to run the integration tests against containerized AGW
  - any obstacles while trying to run the tests against the containerized AGW
  - the test results
- [x] Prepare internal review meeting

## Implementation details
- Service restarts via systemd need to be adapted (?)
- Reference from the integ test run against the deb deployed vm: https://github.com/magma/magma/issues/13263 
- Start the containerized AGW on the current dev VM.

## Additional information
- https://github.com/magma/magma/blob/master/lte/gateway/docker/README.md
- https://github.com/magma/magma/pull/8771

### PR DESCRIPTION
**Pull request by voisey**
## Summary

Works towards getting all integration tests running versus the containerized AGW, starting with `test_attach_detach`. Modifies system restarts to account for Docker. Also modifies the readme to explain the present situation and how to run the test, as well as adding sctpd to the services that need to be stopped before starting the Docker containers.   

This aims to be a non-breaking PR which can be built upon in the future to get further tests running.

## Test Plan

At present, `test_attach_detach` runs successfully if `MAGMA_DEV_MODE` is set to 1 in the `.env` file.

Notes:
- In the future, we may want to use a Docker override file to set the required environment variable instead of modifying the `.env` file.
- Containers can be started following the instructions here: https://github.com/magma/magma/blob/master/lte/gateway/docker/README.md#running-the-containerized-agw-locally-on-the-magma-vm.

### CODE REVIEW

File: lte/gateway/docker/README.md
Comment (nstng) [inline lte/gateway/docker/README.md]: Is this possible after the images are created, i.e., during runtime?

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (nstng) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: is this robust against environments where docker/systemd is not installed?

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (nstng) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: I am surprised that this works, I would expect this to translate to two ssh commands. How does the second command know about the `cd`?

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (nstng) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: why is `sudo` needed here and not for the `docker-compose` restart?

File: lte/gateway/docker/.env
Comment (nstng) [inline lte/gateway/docker/.env]: iiuc this is now defined as an environment variable on the host and in this file - is there a better way to handle this?

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (jheidbrink) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: nitpick: I think `AssertionError` is really just meant to be raised by `assert`. When manually raising an exception there is usually something more specific that fits better. That said, I'm not sure what would be a good one here ;)

File: lte/gateway/docker/README.md
Comment (voisey) [reply lte/gateway/docker/README.md]: We tried various things but nothing seemed to work. For example, you can parse an environment variable as part of a `docker compose` command, but not as part of the deprecated `docker-compose` command we're using. We thought we might try to migrate this to an override file, for example, but seeing as this was just an attempt to get the first test running, we thought this might be okay for the moment.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (sebathomas) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]:
```suggestion
    def detect_init_system(self) -> InitMode:
```
I think it's always good to document the return and input types of new functions.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (sebathomas) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: I think people often use `ValueError` for generic precondition errors like here even if it is meant for something more specific (strictly-speaking). Another option would be a generic `RuntimeError`.

Comment (jheidbrink) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Not sure if this qualifies for `OSError`, otherwise a more general one would be `RuntimeError`

Comment (sebathomas) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: There's also an `EnvironmentError` that's an alias of `OSError` that might fit. On the other hand, `OSError` refers to I/O issues in its docstring and that doesn't really fit...

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (voisey) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Good point, thanks! I've removed the `sudo` here.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (voisey) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Thanks for pointing this out and for the suggestions! To me `EnvironmentError` makes the most intuitive sense but it seems to have been deprecated, so I think I'll just go for `RuntimeError`.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (voisey) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Good point, ta! I've added that now.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (voisey) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: You're right, thanks for spotting this! I've combined the two lines into one command.

File: lte/gateway/docker/.env
Comment (voisey) [reply lte/gateway/docker/.env]: You're right. There may well be a better way. Do you have any thoughts on this @jheidbrink?

Comment (jheidbrink) [reply lte/gateway/docker/.env]: https://docs.docker.com/compose/env-file/ reads "Values present in the environment at runtime always override those defined inside the .env file". That sounds like using the host variable should just work. Haven't tested it yet though.

Comment (jheidbrink) [reply lte/gateway/docker/.env]: I tried it and it seems it doesn't work with `docker-compose`. The docs also mention `docker compose -e`, but again this does not work with `docker-compose`.

**Review by jheidbrink — APPROVED** [review]
I confirmed that I can run `test_attach_detach` on this branch and it succeeds.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (voisey) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Good point! This actually fails on `magma_deb` where Docker is not installed. How's something like this?
```
    def detect_init_system(self) -> InitMode:
        """Detect whether services are running with Docker or systemd."""

        docker_running = False
        systemd_running = False

        try:
            docker_running = "magmad" in self.exec_command_output("docker ps --filter 'name=magmad' --format '{{.Names}}'")
        except:
            logging.info("Docker is not installed")

        try:
            systemd_running = self.exec_command("systemctl is-active magma@magmad") == 0
        except:
            logging.info("systemd is not installed")
        ...
```
Btw, is it possible `systemctl` would not be installed but `systemd` would be?

Comment (jheidbrink) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: `systemctl` should be present when `systemd` is installed, so I think we can safely assume its presence. In the `except` we should be specific which exceptions we want to catch (probably `CalledProcessError`).

Comment (jheidbrink) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py]: Personally I would not initialize the `_running` variables, but add a `docker_running = False` in the except block, but that's a bit subjective, feel free to do it the way you prefer.

File: lte/gateway/docker/README.md
Comment (voisey) [reply lte/gateway/docker/README.md]: I realise you were asking a difference question. I've updates the docs with the sentence below stating that you need to restart the containers.

**Review by sebathomas — APPROVED** [review]

File: lte/gateway/docker/docker-compose.yaml
Comment (nstng) [inline lte/gateway/docker/docker-compose.yaml]: might this line be redundant? is .env used by default?

**Review by nstng — APPROVED** [review]
lgtm (modulo minor comment)

File: lte/gateway/docker/docker-compose.yaml
Comment (voisey) [reply lte/gateway/docker/docker-compose.yaml]: The line was required to set the added environment variable in the containers, however the paths in `.env` were already being used in `docker-compose.yaml` as default. That said, this modification wasn't ideal so it has been replaced by the use of an override file, `docker-compose.dev.yaml`.

File: lte/gateway/docker/docker-compose.dev.yaml
Comment (crasu) [inline lte/gateway/docker/docker-compose.dev.yaml:1]: Instead of using this file you can add MAGMA_DEV_MODE (without any values) to the .env file. This will pass the value of the env variable from the host system to the container.

Comment (crasu) [reply lte/gateway/docker/docker-compose.dev.yaml:1]: Ok this will not work as it does not propagate the variable into the container

**Comment by sebathomas** [comment]
Could someone from @magma/approvers-agw-integtests have a look at this PR? It's the first step for integration tests of the dockerized AGW.

File: lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
Comment (rsarwad) [inline lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:974]: Hi @sebathomas here it looks good as all services have restarted. Do we need similar handling for test case where an individual service have started. Say for test case, test_sctp_shutdown_while_mme_is_stopped.py mme, mobilityd, pipelined services have stopped and started mme service.
Please find the test case here, https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_sctp_shutdown_while_mme_is_stopped.py

Comment (sebathomas) [reply lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:974]: Thanks for looking into this. Yes, I expect more changes will be necessary to run all tests, this will be handled in the follow-up ticket #13684.

**Review by rsarwad — COMMENTED** [review]
S1ap tester changes looks good to me

**Comment by sebathomas** [comment]
> S1ap tester changes looks good to me

Can you approve the PR then? :)

**Review by rsarwad — APPROVED** [review]
s1ap changes LGTM

**Comment by sebathomas** [comment]
Thank you! I think the PR needs a rebase now because of changes to the semantic-pr CI, but apart from that it's ready to merge.

**Comment by rsarwad** [comment]
> Thank you! I think the PR needs a rebase now because of changes to the semantic-pr CI, but apart from that it's ready to merge.
Yes

**Pull request merged**
